### ISSUE TITLE
Requiring ffi-napi from two different locations on disk crashes Node.js v14

### ISSUE THREAD

**Issue by joelpurra** [node-ffi-napi/node-ffi-napi#96]
- Requiring `ffi-napi` from two different locations on disk crashes Node.js v14. The same code works on Node.js v10, v12.
- For example having `package-01` which references `package-02`, when both reference `ffi-napi`.
- Using `npm link` on either `ffi-napi` or `ref-napi` in both `package-xx` works around the issue.
- Steps to reproduce: https://github.com/joelpurra/ffi-napi-issue-nodejs-v14

Am working on a project with several packages which reference each other. The packages are in different directories on disk (not downloaded by `npm install`), and they each have a `node_modules` directory with a copy of `ffi-napi`.

Suspecting that the same issue can arise if `npm install` cannot "cleanly" dedupe `ffi-napi` to a single version on disk (if I understand deduping correctly). Given that [612 repositories, 109 packages](https://github.com/node-ffi-napi/node-ffi-napi/network/dependents?package_id=UGFja2FnZS00MjgxNDU4OQ%3D%3D) rely on `ffi-napi`, which has at least 3 different major versions, this may very well happen.

---

The crash is not, as initially suspected, related to updating to `node-addon-api` v3.0.0 (https://github.com/node-ffi-napi/node-ffi-napi/pull/94#issuecomment-687610353). Instead the perceived "`node-addon-api` fix" was due to using `npm link` to test `ffi-napi` and `ref-napi` with the updated `node-addon-api` version.

**Comment by lingyuj** [comment]
I also encounter the same issue in my project which depends on ffi-napi, ref-napi, ref-struct-napi and ref-array-napi. In Node 14, I have to use a new version of ffi-napi which depends on ref-napi 2.1.2, while the ref-struct-napi and ref-array-napi depend on ref-napi 1.5.2. The same core dump is observed. Any idea to solve it?

**Comment by splitice** [comment]
I can confirm I hit this with nodejs v14.13.0 today

**Comment by dbartle** [comment]
I'm getting a similar and consistent crash on v14.13.0, with no issues on v12.

We have two packages using FFI and I _believe_ that I only encountered the crash after the second package was completed, which seems like it connects with the original post. I've reproduced it on macOS and Raspberry Pi. 

Using these versions in both projects:
```json
  "dependencies": {
    "ffi-napi": "^3.0.1",
    "ref-array-napi": "^1.2.1",
    "ref-napi": "^3.0.1"
  },
```

**Comment by lingyuj** [comment]
This is a known issue of ffi-napi. We have a workaround for this issue in the next release. Currently please use an older node.js version or if you have to use node.js v14.13, please let us know.

**Comment by samuelraeburn** [comment]
@lingyuj When is the next release scheduled for?

**Comment by lingyuj** [comment]
> @lingyuj When is the next release scheduled for?

Sorry I replied the wrong thread. I am not the owner of this module.

**Comment by joelpurra** [comment]
@addaleax: any progress on this, upstream or otherwise?

**Comment by addaleax** [comment]
@joelpurra No, not yet, and thank you for pinging me about this. I’m honestly not sure what a good solution for this would look like. It might be possible to coordinate this for the case where the respective versions of ffi-napi and ref-napi are equal, but I’m not sure what it would look like in the alternative case.

**Comment by hunter-sd** [comment]
> The issue seems to be the usage of a `Buffer<char>` in `WrapPointer` to pass a C ptr to node (and get it back later in callbacks). This is an incorrect usage of NAPI, an `External<T>::New(env,ptr)` should be used instead, although it's then not an actual object, as the Buffer with 0 size seems to create some GC issue with Node 14.16
> 
> [edit] upgrading to ffi-napi 4.x fixed it

I am currently using node 14.16.1 and ffi-napi 4.0.3 and am still seeing the issue. Any other ideas on a fix for this?

**Comment by stavshamir** [comment]
I am also using node 14.16.1 with ffi-napi 4.03. and seeing this issue.
Thanks @hunter-sd, your workaround works well.

**Comment by creasman** [comment]
We hit this issue through a package dependency.  The package was installing ffi-napi, version 3.x.  Once we upgraded the package so that it used ffi-napi, version 4.0.3, the problem was resolved.
